vantage-agent/lib/charms/vantage_agent/v0/agent_snapper.py

```diff
@@ -0,0 +1,193 @@
+            self._charm.on.config_changed: self.handle_config_changed,
+            self._charm.on.stop: self.handle_stop,
+            self._charm.on.start: self.handle_start,
+            self._charm.on.remove: self.handle_remove,
+        }
+        for event, handler in event_handler_bindings.items():
+            self._charm.framework.observe(event, handler)
+
+    ## Event Handlers
+    def handle_install(self, event: InstallEvent):
+        logger.debug(f"## Processing install event for {self._snap_name}.")
+        self._charm.unit.status = WaitingStatus(f"Installing snap for {self._snap_name}")
+        try:
+            self.install_snap()
+        except Exception as e:
+            logger.error(f"## Error installing {self._snap_name}: {e}")
+            self._charm.unit.status = BlockedStatus("Error installing the snap for {self._snap_name}")
+            event.defer()
+            return
+
+        logger.debug(f"Snap for {self._snap_name} installed")
+
+    def handle_config_changed(self, event: ConfigChangedEvent):
+        logger.debug(f"## Processing config changed event for {self._snap_name}.")
+        if not self.is_snap_installed():
+            logger.debug(f"## Snap for {self._snap_name} not installed, deferring event")
+            event.defer()
+            return
+
+        logger.info(f"## Updating config for {self._snap_name}")
+        self.configure_snap()
+
+        if self.is_snap_active():
+            self.run_snap_service("restart")
+
+    def handle_start(self, event: StopEvent):
+        logger.debug(f"## Processing start event for {self._snap_name}.")
+        self.run_snap_service("start")
+        self._charm.unit.status = ActiveStatus()
+
+    def handle_stop(self, event: StopEvent):
+        logger.debug(f"## Processing stop event for {self._snap_name}.")
+        self.run_snap_service("stop")
+        self._charm.unit.status = MaintenanceStatus()
+
+    def handle_remove(self, event: RemoveEvent):
+        logger.debug(f"## Processing remove event for {self._snap_name}.")
+        self.remove_snap()
+
+    ## Operations
+    def is_snap_installed(self) -> bool:
+        cmd = [
+            self._snap_path,
+            "list",
+            self._snap_name,
+        ]
+
+        try:
+            subprocess.check_output(cmd)
+            return True
+        except subprocess.CalledProcessError:
+            return False
+
+    def is_snap_active(self) -> bool:
+        logger.debug(f"### Checking active status for {self._snap_name}.daemon")
+        cmd = [
+            self._snap_path,
+            "services",
+            f"{self._snap_name}.daemon",
+        ]
+
+        try:
+            status_output = subprocess.check_output(cmd).decode("utf-8")
+            status_line = status_output.split("\n")[1]
+            status = status_line.split()[2]
+            return status == "active"
+        except subprocess.CalledProcessError:
+            return False
```

Comment:
I think it'd be better to run the command `snap services <snap name>.daemon | tail -n1 | awk '{print $3}'` and get directly the `active` or `inactive` state than parsing the output as is. The reason for that is that there might be cases where the snap isn't installed for example and other errors might be raised by the python logic, while running the command would fail and the python code would capture the `CalledProcessError` properly.

Reply:
You can't use pipes in subprocess args. You have to actually call the commands separately and connect them with a pipe.

Instead of doing that, I'll update this logic to be more robust.

Reply:
Resolved in commit e15fece.

Reply:
very nice implementation